### PR TITLE
Fix 20916 - Print trace for deprecations triggered inside templates

### PR DESCRIPTION
**Pull request by Geod24**
This turned out to be way easier than I expected.
@John-Colvin : Want to give this a shot ?

I also think it segv dtoh (and I have an idea where the bug is), but let's see what the auto-tester says...

### CODE REVIEW

**Comment by Geod24** [comment]
@ibuclaw @kinke : Do you need to access `printInstantiationTrace` from C++ ? If not, I'll make it `extern(D)` and `final`.
The issue is that `Classification` uses `Color` (which is an `enum` of `int`). [dtoh does not like the way `Classification` is defined](https://issues.dlang.org/show_bug.cgi?id=21300).

**Comment by kinke** [comment]
I'm almost certain LDC doesn't.

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -5984,6 +5984,10 @@ extern (C++) class TemplateInstance : ScopeDsymbol
         for (TemplateInstance cur = this; cur; cur = cur.tinst)
         {
             ++n_instantiations;
+            // Set error here as we don't want it to depend on the number of
+            // entries that are being printed.
+            cur.errors = true;
```
Comment (MoonlightSentinel) [inline src/dmd/dtemplate.d]: What caused this commit?

Comment (Geod24) [reply src/dmd/dtemplate.d]: It just seems very wrong to have this set *only* on the templates we show (which can vary depending on `-v`).
`errors` has meaning semantic-wise, and the amount of frames we print should not affect the semantic of the code.
I didn't witness anything wrong that prompted the commit, it's just my observation of what *should* happen.
Note that I believe the diff on `test/fail_compilation/deprecations.d` (L9 in the old code) is due to this: We don't show the error message twice anymore.

**Comment by Geod24** [comment]
This should be good to go (there was a spurious Buildkite failure, re-triggered).

**Review by thewilsonator — APPROVED** [review]

**Comment by ibuclaw** [comment]
Well this pulls in excessive baggage to the compiler front-end.
```
dmd/dtemplate.d:5972:26: error: undefined identifier ‘Classification’
 5972 |     extern(D) final void printInstantiationTrace(Classification cl = Classification.error)
      |                          ^
```
Copying over Classification from dmd ad verbatim, but now there's a missing module.
```
dmd/errors.d:24:13: error: undefined identifier ‘Color’
   24 |     error = Color.brightRed,          /// for errors
```
This occurs in a non-shareable module, so not a big deal, but just highlighting that maybe an another enum would be better.

**Pull request merged**